File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/mail_detail/variants/CouponDetailLayout.kt
@file:Suppress(
    "PackageNaming",
    "MagicNumber",
    "LongMethod",
    "LongParameterList",
    "TooManyFunctions",
    "ComplexMethod",
)

package app.pantopus.android.ui.screens.mailbox.mail_detail.variants

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.data.api.models.mailbox.v2.CouponDetailDto
import app.pantopus.android.ui.screens.mailbox.item_detail.MailItemCategory
import app.pantopus.android.ui.screens.mailbox.item_detail.bodies.CouponBody
import app.pantopus.android.ui.screens.mailbox.item_detail.bodies.CouponBodyState
import app.pantopus.android.ui.screens.mailbox.mail_detail.MailDetailContent
import app.pantopus.android.ui.screens.mailbox.mail_detail.MailDetailKeyFact
import app.pantopus.android.ui.screens.shared.mail_item_detail.AIElfStripContent
import app.pantopus.android.ui.screens.shared.mail_item_detail.AttachmentItem
import app.pantopus.android.ui.screens.shared.mail_item_detail.AttachmentKind
import app.pantopus.android.ui.screens.shared.mail_item_detail.AttachmentsRowContent
import app.pantopus.android.ui.screens.shared.mail_item_detail.MailItemDetailShell
import app.pantopus.android.ui.screens.shared.mail_item_detail.MailOverflowItem
import app.pantopus.android.ui.screens.shared.mail_item_detail.MailTopBarConfig
import app.pantopus.android.ui.screens.shared.mail_item_detail.MailTopBarTrailingAction
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing
import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter

/**
 * A17.5 — Coupon ceremonial variant of the mail item detail. Mirrors
 * iOS `CouponDetailLayout`. Body slot is the existing [CouponBody]
 * (ticket hero + fine-print + barcode-expand card). The actions shelf
 * hosts the redemption-state CTA: Redeem primary while live, an
 * "Already redeemed" pill once flipped, and a terminal "Expired"
 * affordance.
 *
 * [today] is injectable so callers that must be deterministic — snapshot
 * tests in particular — can pin the expiry comparison instead of reading
 * the ambient clock. Production callers take the default.
 */
@Composable
fun CouponDetailLayout(
    content: MailDetailContent,
    coupon: CouponDetailDto,
    redeemInFlight: Boolean,
    onBack: () -> Unit,
    onRedeem: () -> Unit,
    onOpenSenderProfile: (String) -> Unit = {},
    onSaveToVault: () -> Unit = {},
    today: LocalDate = LocalDate.now(),
) {
    val state = bodyState(content, coupon, today)
    Box(modifier = Modifier.testTag("mailDetail_coupon")) {
        MailItemDetailShell(
            topBar = makeTopBar(content = content, onBack = onBack, onSaveToVault = onSaveToVault),
            aiElf = makeAIElf(state = state, coupon = coupon),
            attachments = makeAttachments(content = content),
            hero = { CouponHeroCard(content = content) },
            keyFacts = { CouponKeyFactsCard(rows = makeKeyFacts(coupon = coupon)) },
            body = { CouponBody(coupon = coupon, state = state) },
            sender = { SenderCard(content = content, onOpenProfile = onOpenSenderProfile) },
            actions = {
                CouponActionsRow(
                    state = state,
                    redeemInFlight = redeemInFlight,
                    onRedeem = onRedeem,
                    onSaveToVault = onSaveToVault,
                )
            },
        )
    }
}

private fun bodyState(
    content: MailDetailContent,
    coupon: CouponDetailDto,
    today: LocalDate,
): CouponBodyState {
    if (content.isAcknowledged) return CouponBodyState.Redeemed
    val raw = coupon.expiresAt?.takeIf { it.isNotBlank() } ?: return CouponBodyState.Unused
    val parsed =
        runCatching { OffsetDateTime.parse(raw).toLocalDate() }.getOrNull()
            ?: runCatching { LocalDate.parse(raw) }.getOrNull()
            ?: runCatching { LocalDate.parse(raw, DateTimeFormatter.ofPattern("yyyy-MM-dd")) }.getOrNull()
    return if (parsed != null && parsed.isBefore(today)) {
        CouponBodyState.Expired
    } else {
        CouponBodyState.Unused
    }
}

private fun makeTopBar(
    content: MailDetailContent,
    onBack: () -> Unit,
    onSaveToVault: () -> Unit,
): MailTopBarConfig =
    MailTopBarConfig(
        eyebrow = "Coupon",
        trust = content.detailTrust,
        onBack = onBack,
        trailingAction =
            MailTopBarTrailingAction(
                icon = PantopusIcon.Bookmark,
                contentDescription = "Save to vault",
                onClick = onSaveToVault,
            ),
        overflowItems =
            listOf(
                MailOverflowItem("share", PantopusIcon.Share, "Share") {},
                MailOverflowItem("saveToVault", PantopusIcon.Bookmark, "Save to vault") { onSaveToVault() },
                MailOverflowItem("addToWallet", PantopusIcon.Wallet, "Add to wallet") {},
                MailOverflowItem("archive", PantopusIcon.Archive, "Archive") {},
                MailOverflowItem(
                    id = "delete",
                    icon = PantopusIcon.Trash2,
                    label = "Delete",
                    isDestructive = true,
                ) {},
            ),
    )

private fun makeAIElf(
    state: CouponBodyState,
    coupon: CouponDetailDto,
): AIElfStripContent {
    val merchant = coupon.merchant ?: coupon.brandName ?: "the merchant"
    val (headline, summary) =
        when (state) {
            CouponBodyState.Unused ->
                "Saved for your next visit" to
                    "Show this at checkout — Pantopus will mute reminders once you redeem."
            CouponBodyState.Redeemed ->
                "Redeemed at $merchant" to
                    "The single-use barcode is retired. We'll keep a copy in your Vault for receipts."
            CouponBodyState.Expired ->
                "This offer has expired" to
                    "You can archive this mail or move on — the barcode is no longer scannable."
        }
    return AIElfStripContent(headline = headline, summary = summary)
}

private fun makeAttachments(content: MailDetailContent): AttachmentsRowContent? {
    if (content.attachments.isEmpty()) return null
    val items =
        content.attachments.mapIndexed { index, name ->
            AttachmentItem(id = "att-$index", kind = AttachmentKind.Pdf, name = name)
        }
    return AttachmentsRowContent(title = "Fine print", items = items)
}

private fun makeKeyFacts(coupon: CouponDetailDto): List<MailDetailKeyFact> =
    buildList {
        (coupon.merchant ?: coupon.brandName)?.let {
            add(MailDetailKeyFact(icon = PantopusIcon.Briefcase, label = "Merchant", value = it))
        }
        coupon.code?.let {
            add(MailDetailKeyFact(icon = PantopusIcon.Hash, label = "Code", value = it))
        }
        coupon.minimumSpend?.let {
            add(MailDetailKeyFact(icon = PantopusIcon.Info, label = "Min. spend", value = it))
        }
        coupon.expiresAt?.let {
            add(MailDetailKeyFact(icon = PantopusIcon.Clock, label = "Expires", value = it))
        }
    }

@Composable
private fun CouponHeroCard(content: MailDetailContent) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg)),
    ) {
        Box(
            modifier =
                Modifier
                    .width(4.dp)
                    .fillMaxHeight()
                    .background(content.category.accent),
        )
        Column(
            modifier = Modifier.padding(Spacing.s3),
            verticalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                CategoryBadge(category = content.category)
                Spacer(Modifier.weight(1f))
                content.createdAtLabel?.let { received ->
                    Text(text = received, fontSize = 11.sp, color = PantopusColors.appTextSecondary)
                }
            }
            Text(
                text = content.senderDisplayName.uppercase(),
                fontSize = 11.sp,
                fontWeight = FontWeight.SemiBold,
                letterSpacing = 0.6.sp,
                color = PantopusColors.appTextSecondary,
            )
            Text(
                text = content.title,
                fontSize = 19.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appText,
                lineHeight = 24.sp,
                modifier = Modifier.semantics { heading() },
            )
            content.excerpt?.takeIf { it.isNotEmpty() }?.let {
                Text(
                    text = it,
                    fontSize = 12.sp,
                    color = PantopusColors.appTextSecondary,
                )
            }
        }
    }
}

@Composable
private fun CategoryBadge(category: MailItemCategory) {
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(category.rowBackground)
                .padding(horizontal = Spacing.s2, vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        PantopusIconImage(
            icon = category.icon,
            contentDescription = null,
            size = 11.dp,
            tint = category.accent,
        )
        Text(
            text = category.label,
            fontSize = 10.sp,
            fontWeight = FontWeight.Bold,
            letterSpacing = 0.4.sp,
            color = category.accent,
        )
    }
}

@Composable
private fun CouponKeyFactsCard(rows: List<MailDetailKeyFact>) {
    if (rows.isEmpty()) return
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg)),
    ) {
        Text(
            text = "OFFER FACTS",
            modifier =
                Modifier
                    .padding(horizontal = Spacing.s3, vertical = Spacing.s2)
                    .semantics { heading() },
            fontSize = 11.sp,
            fontWeight = FontWeight.Bold,
            letterSpacing = 0.5.sp,
            color = PantopusColors.appTextSecondary,
        )
        HorizontalDivider(color = PantopusColors.appBorderSubtle)
        rows.forEachIndexed { index, row ->
            Row(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .padding(horizontal = Spacing.s3, vertical = Spacing.s2),
                verticalAlignment = Alignment.Top,
                horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
            ) {
                Box(
                    modifier =
                        Modifier
                            .size(24.dp)
                            .clip(RoundedCornerShape(Radii.sm))
                            .background(PantopusColors.appSurfaceSunken),
                    contentAlignment = Alignment.Center,
                ) {
                    PantopusIconImage(
                        icon = row.icon,
                        contentDescription = null,
                        size = 13.dp,
                        tint = PantopusColors.appTextStrong,
                    )
                }
                Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(1.dp)) {
                    Text(
                        text = row.label.uppercase(),
                        fontSize = 11.sp,
                        fontWeight = FontWeight.SemiBold,
                        letterSpacing = 0.4.sp,
                        color = PantopusColors.appTextSecondary,
                    )
                    Text(
                        text = row.value,
                        fontSize = 13.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = PantopusColors.appText,
                    )
                }
            }
            if (index < rows.size - 1) {
                HorizontalDivider(color = PantopusColors.appBorderSubtle)
            }
        }
    }
}

@Composable
private fun SenderCard(
    content: MailDetailContent,
    onOpenProfile: (String) -> Unit,
) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                .clickable(enabled = content.senderUserId != null) {
                    content.senderUserId?.let(onOpenProfile)
                }
                .padding(Spacing.s3),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Box(
            modifier =
                Modifier
                    .size(44.dp)
                    .clip(RoundedCornerShape(Radii.lg))
                    .background(content.category.accent),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = content.senderInitials,
                fontSize = 14.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appTextInverse,
            )
        }
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(
                text = content.senderDisplayName,
                fontSize = 14.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appText,
            )
            content.senderMeta?.let {
                Text(text = it, fontSize = 12.sp, color = PantopusColors.appTextSecondary)
            }
        }
        if (content.senderUserId != null) {
            PantopusIconImage(
                icon = PantopusIcon.ChevronRight,
                contentDescription = null,
                size = 14.dp,
                tint = PantopusColors.appTextMuted,
            )
        }
    }
}

@Composable
private fun CouponActionsRow(
    state: CouponBodyState,
    redeemInFlight: Boolean,
    onRedeem: () -> Unit,
    onSaveToVault: () -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(Spacing.s2)) {
        when (state) {
            CouponBodyState.Unused ->
                RedeemButton(redeemInFlight = redeemInFlight, onRedeem = onRedeem)
            CouponBodyState.Redeemed -> RedeemedPill()
            CouponBodyState.Expired -> ExpiredPill()
        }
        Row(horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
            SecondaryTile(
                id = "save",
                icon = PantopusIcon.Bookmark,
                label = "Save",
                onClick = onSaveToVault,
                modifier = Modifier.weight(1f),
            )
            SecondaryTile(
                id = "share",
                icon = PantopusIcon.Share,
                label = "Share",
                modifier = Modifier.weight(1f),
            )
            SecondaryTile(
                id = "directions",
                icon = PantopusIcon.MapPin,
                label = "Get directions",
                modifier = Modifier.weight(1f),
            )
        }
    }
}

@Composable
private fun RedeemButton(
    redeemInFlight: Boolean,
    onRedeem: () -> Unit,
) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(14.dp))
                .background(PantopusColors.primary600)
                .clickable(enabled = !redeemInFlight, onClick = onRedeem)
                .padding(vertical = 14.dp)
                .alpha(if (redeemInFlight) 0.6f else 1f)
                .semantics { contentDescription = "Mark redeemed" }
                .testTag("mailDetail_coupon_redeem"),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center,
    ) {
        PantopusIconImage(
            icon = PantopusIcon.CheckCircle,
            contentDescription = null,
            size = 16.dp,
            tint = PantopusColors.appTextInverse,
        )
        Spacer(Modifier.width(Spacing.s2))
        Text(
            text = "Mark redeemed",
            fontSize = 15.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.appTextInverse,
        )
    }
}

@Composable
private fun RedeemedPill() {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(14.dp))
                .background(PantopusColors.successBg)
                .border(1.5.dp, PantopusColors.successLight, RoundedCornerShape(14.dp))
                .padding(vertical = 14.dp)
                .testTag("mailDetail_coupon_redeemed"),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center,
    ) {
        PantopusIconImage(
            icon = PantopusIcon.CheckCircle,
            contentDescription = null,
            size = 16.dp,
            tint = PantopusColors.success,
        )
        Spacer(Modifier.width(Spacing.s2))
        Text(
            text = "Already redeemed",
            fontSize = 15.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.success,
        )
    }
}

@Composable
private fun ExpiredPill() {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(14.dp))
                .background(PantopusColors.errorBg)
                .border(1.5.dp, PantopusColors.error.copy(alpha = 0.2f), RoundedCornerShape(14.dp))
                .padding(vertical = 14.dp)
                .testTag("mailDetail_coupon_expired"),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center,
    ) {
        PantopusIconImage(
            icon = PantopusIcon.AlertCircle,
            contentDescription = null,
            size = 16.dp,
            tint = PantopusColors.error,
        )
        Spacer(Modifier.width(Spacing.s2))
        Text(
            text = "This offer has expired",
            fontSize = 15.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.error,
        )
    }
}

@Composable
private fun SecondaryTile(
    id: String,
    icon: PantopusIcon,
    label: String,
    onClick: () -> Unit = {},
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                .clickable(onClick = onClick)
                .padding(vertical = 10.dp)
                .semantics { contentDescription = label }
                .testTag("mailDetail_coupon_action_$id"),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        PantopusIconImage(
            icon = icon,
            contentDescription = null,
            size = 17.dp,
            tint = PantopusColors.appTextStrong,
        )
        Text(
            text = label,
            fontSize = 10.5.sp,
            fontWeight = FontWeight.SemiBold,
            color = PantopusColors.appTextStrong,
        )
    }
}

File: frontend/apps/ios/Pantopus/Features/Mailbox/MailDetail/Variants/CouponDetailLayout.swift
//
//  CouponDetailLayout.swift
//  Pantopus
//
//  A17.5 — Coupon ceremonial variant of the mail item detail. Sits on
//  the shared `MailItemDetailShell` (P19); the body slot is the
//  bespoke `CouponBody` (ticket hero + fine-print + barcode-expand
//  card). The action shelf hosts the redemption-state CTA: Redeem
//  primary while the offer is live, an "Already redeemed" disabled
//  pill once flipped, and a terminal "Expired" affordance.
//

import SwiftUI

// swiftlint:disable multiple_closures_with_trailing_closure

@MainActor
struct CouponDetailLayout: View {
    let content: MailDetailContent
    let coupon: CouponDetailDTO
    let redeemInFlight: Bool
    let onBack: @MainActor () -> Void
    let onRedeem: @MainActor () -> Void
    let onOpenSenderProfile: (@MainActor (String) -> Void)?
    var onSaveToVault: @MainActor () -> Void = {}

    var body: some View {
        MailItemDetailShell(
            topBar: makeTopBar(),
            aiElf: makeAIElf(),
            attachments: makeAttachments(),
            hero: { CouponHeroCard(content: content) },
            keyFacts: { KeyFactsCard(rows: makeKeyFacts()) },
            body: { CouponBody(coupon: coupon, state: bodyState) },
            sender: { SenderCard(content: content, onOpenProfile: onOpenSenderProfile) },
            actions: {
                CouponActionsRow(
                    state: bodyState,
                    redeemInFlight: redeemInFlight,
                    onRedeem: onRedeem,
                    onSaveToVault: onSaveToVault
                )
            }
        )
        .accessibilityIdentifier("mailDetail_coupon")
    }

    private var bodyState: CouponBodyState {
        if content.isAcknowledged { return .redeemed }
        // Best-effort expiry check — backend ships `expires_at` as ISO 8601
        // or YYYY-MM-DD; either form parses through the standard
        // formatter. Falls open to `.unused` when we can't parse it.
        if let raw = coupon.expiresAt, let date = parseExpiry(raw), date < Date() {
            return .expired
        }
        return .unused
    }

    private func parseExpiry(_ raw: String) -> Date? {
        let iso = ISO8601DateFormatter()
        if let date = iso.date(from: raw) { return date }
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.date(from: raw)
    }

    private func makeTopBar() -> MailTopBarConfig {
        MailTopBarConfig(
            eyebrow: "Coupon",
            trust: content.detailTrust,
            onBack: { @Sendable in Task { @MainActor in onBack() } },
            trailingAction: MailTopBarTrailingAction(
                icon: .bookmark,
                accessibilityLabel: "Save to vault"
            ) { @Sendable in Task { @MainActor in onSaveToVault() } },
            overflowItems: [
                MailOverflowItem(id: "share", icon: .share, label: "Share") {},
                MailOverflowItem(id: "saveToVault", icon: .bookmark, label: "Save to vault") { @Sendable in
                    Task { @MainActor in onSaveToVault() }
                },
                MailOverflowItem(id: "addToWallet", icon: .wallet, label: "Add to wallet") {},
                MailOverflowItem(id: "archive", icon: .archive, label: "Archive") {},
                MailOverflowItem(id: "delete", icon: .trash2, label: "Delete", isDestructive: true) {}
            ]
        )
    }

    private func makeAIElf() -> AIElfStripContent? {
        let merchant = coupon.merchant ?? coupon.brandName ?? "the merchant"
        let headline: String
        let summary: String
        switch bodyState {
        case .unused:
            headline = "Saved for your next visit"
            summary = "Show this at checkout — Pantopus will mute reminders once you redeem."
        case .redeemed:
            headline = "Redeemed at \(merchant)"
            summary = "The single-use barcode is retired. We'll keep a copy in your Vault for receipts."
        case .expired:
            headline = "This offer has expired"
            summary = "You can archive this mail or move on — the barcode is no longer scannable."
        }
        return AIElfStripContent(headline: headline, summary: summary)
    }

    private func makeAttachments() -> AttachmentsRowContent? {
        guard !content.attachments.isEmpty else { return nil }
        let items = content.attachments.enumerated().map { index, name in
            AttachmentItem(id: "att-\(index)", kind: .pdf, name: name)
        }
        return AttachmentsRowContent(title: "Fine print", items: items)
    }

    private func makeKeyFacts() -> [MailDetailKeyFact] {
        var rows: [MailDetailKeyFact] = []
        if let merchant = coupon.merchant ?? coupon.brandName {
            rows.append(MailDetailKeyFact(icon: .briefcase, label: "Merchant", value: merchant))
        }
        if let code = coupon.code {
            rows.append(MailDetailKeyFact(icon: .hash, label: "Code", value: code))
        }
        if let minimumSpend = coupon.minimumSpend {
            rows.append(MailDetailKeyFact(icon: .info, label: "Min. spend", value: minimumSpend))
        }
        if let expiresAt = coupon.expiresAt {
            rows.append(MailDetailKeyFact(icon: .clock, label: "Expires", value: expiresAt))
        }
        return rows
    }
}

// MARK: - Hero

private struct CouponHeroCard: View {
    let content: MailDetailContent

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            HStack(spacing: Spacing.s1) {
                CategoryBadge(category: content.category)
                Spacer()
                if let received = content.createdAtLabel {
                    Text(received)
                        .font(.system(size: 11))
                        .foregroundStyle(Theme.Color.appTextSecondary)
                }
            }
            Text(content.senderDisplayName.uppercased())
                .font(.system(size: 11, weight: .semibold))
                .tracking(0.6)
                .foregroundStyle(Theme.Color.appTextSecondary)
            Text(content.title)
                .font(.system(size: 19, weight: .bold))
                .foregroundStyle(Theme.Color.appText)
                .fixedSize(horizontal: false, vertical: true)
            if let excerpt = content.excerpt, !excerpt.isEmpty {
                Text(excerpt)
                    .pantopusTextStyle(.caption)
                    .foregroundStyle(Theme.Color.appTextSecondary)
            }
        }
        .padding(Spacing.s3)
        .background(Theme.Color.appSurface)
        .overlay(alignment: .leading) {
            Rectangle().fill(content.category.accent).frame(width: 4)
        }
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
    }
}

private struct CategoryBadge: View {
    let category: MailItemCategory

    var body: some View {
        HStack(spacing: Spacing.s1) {
            Icon(category.icon, size: 11, color: category.accent)
            Text(category.label)
                .font(.system(size: 10, weight: .bold))
                .tracking(0.4)
                .foregroundStyle(category.accent)
        }
        .padding(.horizontal, Spacing.s2)
        .padding(.vertical, 3)
        .background(category.rowBackground)
        .clipShape(RoundedRectangle(cornerRadius: Radii.pill))
    }
}

// MARK: - Key facts

private struct KeyFactsCard: View {
    let rows: [MailDetailKeyFact]

    var body: some View {
        if rows.isEmpty {
            EmptyView()
        } else {
            VStack(alignment: .leading, spacing: Spacing.s0) {
                Text("OFFER FACTS")
                    .font(.system(size: 11, weight: .bold))
                    .tracking(0.5)
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, Spacing.s3)
                    .padding(.vertical, Spacing.s2)
                Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
                ForEach(Array(rows.enumerated()), id: \.element.id) { index, row in
                    HStack(alignment: .top, spacing: Spacing.s3) {
                        Icon(row.icon, size: 13, color: Theme.Color.appTextStrong)
                            .frame(width: 24, height: 24)
                            .background(Theme.Color.appSurfaceSunken)
                            .clipShape(RoundedRectangle(cornerRadius: Radii.sm))
                        VStack(alignment: .leading, spacing: 1) {
                            Text(row.label.uppercased())
                                .font(.system(size: 11, weight: .semibold))
                                .tracking(0.4)
                                .foregroundStyle(Theme.Color.appTextSecondary)
                            Text(row.value)
                                .font(.system(size: 13, weight: .semibold))
                                .foregroundStyle(Theme.Color.appText)
                        }
                        Spacer(minLength: Spacing.s0)
                    }
                    .padding(.horizontal, Spacing.s3)
                    .padding(.vertical, Spacing.s2)
                    if index < rows.count - 1 {
                        Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
                    }
                }
            }
            .background(Theme.Color.appSurface)
            .overlay(
                RoundedRectangle(cornerRadius: Radii.lg)
                    .stroke(Theme.Color.appBorder, lineWidth: 1)
            )
            .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
        }
    }
}

// MARK: - Sender

private struct SenderCard: View {
    let content: MailDetailContent
    let onOpenProfile: (@MainActor (String) -> Void)?

    var body: some View {
        HStack(spacing: Spacing.s3) {
            Text(content.senderInitials)
                .font(.system(size: 14, weight: .bold))
                .foregroundStyle(Theme.Color.appTextInverse)
                .frame(width: 44, height: 44)
                .background(content.category.accent)
                .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
            VStack(alignment: .leading, spacing: 2) {
                Text(content.senderDisplayName)
                    .font(.system(size: 14, weight: .bold))
                    .foregroundStyle(Theme.Color.appText)
                if let meta = content.senderMeta {
                    Text(meta)
                        .pantopusTextStyle(.caption)
                        .foregroundStyle(Theme.Color.appTextSecondary)
                }
            }
            Spacer(minLength: Spacing.s0)
            if onOpenProfile != nil, content.senderUserId != nil {
                Icon(.chevronRight, size: 14, color: Theme.Color.appTextMuted)
            }
        }
        .padding(Spacing.s3)
        .contentShape(Rectangle())
        .onTapGesture {
            if let onOpenProfile, let userId = content.senderUserId {
                onOpenProfile(userId)
            }
        }
        .background(Theme.Color.appSurface)
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
    }
}

// MARK: - Redemption-state actions

private struct CouponActionsRow: View {
    let state: CouponBodyState
    let redeemInFlight: Bool
    let onRedeem: @MainActor () -> Void
    let onSaveToVault: @MainActor () -> Void

    var body: some View {
        VStack(spacing: Spacing.s2) {
            primaryButton
            HStack(spacing: Spacing.s2) {
                secondary(id: "save", icon: .bookmark, label: "Save", action: onSaveToVault)
                secondary(id: "share", icon: .share, label: "Share")
                secondary(id: "directions", icon: .mapPin, label: "Get directions")
            }
        }
    }

    @ViewBuilder
    private var primaryButton: some View {
        switch state {
        case .unused:
            redeemButton
        case .redeemed:
            redeemedPill
        case .expired:
            expiredPill
        }
    }

    private var redeemButton: some View {
        Button(action: { onRedeem() }) {
            HStack(spacing: Spacing.s2) {
                Icon(.checkCircle, size: 16, color: Theme.Color.appTextInverse)
                Text("Mark redeemed")
                    .font(.system(size: 15, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextInverse)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 14)
            .background(Theme.Color.primary600)
            .clipShape(RoundedRectangle(cornerRadius: 14))
            .opacity(redeemInFlight ? 0.6 : 1)
        }
        .buttonStyle(.plain)
        .disabled(redeemInFlight)
        .accessibilityIdentifier("mailDetail_coupon_redeem")
    }

    private var redeemedPill: some View {
        HStack(spacing: Spacing.s2) {
            Icon(.checkCircle, size: 16, color: Theme.Color.success)
            Text("Already redeemed")
                .font(.system(size: 15, weight: .bold))
                .foregroundColor(Theme.Color.success)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 14)
        .background(Theme.Color.successBg)
        .overlay(
            RoundedRectangle(cornerRadius: 14)
                .stroke(Theme.Color.successLight, lineWidth: 1.5)
        )
        .clipShape(RoundedRectangle(cornerRadius: 14))
        .accessibilityIdentifier("mailDetail_coupon_redeemed")
    }

    private var expiredPill: some View {
        HStack(spacing: Spacing.s2) {
            Icon(.alertCircle, size: 16, color: Theme.Color.error)
            Text("This offer has expired")
                .font(.system(size: 15, weight: .bold))
                .foregroundColor(Theme.Color.error)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 14)
        .background(Theme.Color.errorBg)
        .overlay(
            RoundedRectangle(cornerRadius: 14)
                .stroke(Theme.Color.error.opacity(0.2), lineWidth: 1.5)
        )
        .clipShape(RoundedRectangle(cornerRadius: 14))
        .accessibilityIdentifier("mailDetail_coupon_expired")
    }

    private func secondary(
        id: String,
        icon: PantopusIcon,
        label: String,
        action: @escaping @MainActor () -> Void = {}
    ) -> some View {
        Button(action: { action() }) {
            VStack(spacing: Spacing.s1) {
                Icon(icon, size: 17, color: Theme.Color.appTextStrong)
                Text(label)
                    .font(.system(size: 10.5, weight: .semibold))
                    .foregroundStyle(Theme.Color.appTextStrong)
                    .lineLimit(1)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 10)
            .background(Theme.Color.appSurface)
            .overlay(
                RoundedRectangle(cornerRadius: Radii.lg)
                    .stroke(Theme.Color.appBorder, lineWidth: 1)
            )
            .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("mailDetail_coupon_action_\(id)")
    }
}
